### PR TITLE
Close validation gaps in project validators

### LINKED ISSUE #1051: Validation gaps: parameter-set validator never runs, missing cross-reference checks, cli glue injection
Findings from a full-codebase audit of the chapter 8d branch (`excel-deprecation-and-reorg` at b08c6d57, 2026-06-10). File and line references are against that branch. Part of #1030. The two headline findings were reproduced by execution.

`validateProject()` is the gate that `runScenarios()` / `createPlots()` / `runPI()` rely on, but it has holes exactly where the chapter seams are: things it waves through then fail later with worse errors, or never fail at all.

## Findings

- [x] **Critical**, R/validation.R:435-453: `.validateParameterSets()` reads `set$paths` / `set$values` / `set$units`, a shape neither the parser nor the mutators produce (sets are lists of `{containerPath, parameterName, value, units}` records), so `set$paths` is always NULL and every set short-circuits at `length(paths) == 0`. **All three parameter-set sections are effectively unvalidated.** Reproduced: a set containing `list(containerPath = "", parameterName = "", value = "abc", units = NULL)` passes with zero critical errors and zero warnings. The only existing test (test-validation.R:292-299) feeds the legacy shape, which is why this never fired. Related to #1026 (validators should operate on the JSON shapes directly); fixing this validator is the concrete first step.
- [x] R/validation.R:255-273: `.abortValidationErrors()` passes raw validation messages (which embed user-controlled ids) into `cli::cli_abort()` bullets where they are glue-interpolated. A scenario named `Dose {mg}` makes the abort itself crash ("Could not evaluate cli `{}` expression"); `S{1}` silently renders as `S1`. Escape braces before interpolation.
- [x] Scenario to Output Path is the one inbound reference with **no cross-reference check**: it is in `.warnIfReferenced()`'s entity list (R/validation.R:342-344) and the docstring (280-282) promises the next `validateProject()` will surface dangling references, but no check exists. Reproduced end to end: `removeOutputPath()` of a path a scenario uses, then `validateProject()` reports zero critical errors, then `saveProject()` aborts in `.scenariosToJson()` (R/project-to-json.R:140-145).
- [x] dataCombined `observed[*]$dataSet` references are never checked (R/validation.R:662-682 checks only `simulated`), and `removeObservedData()` is the only remover in the package without a referenced-elsewhere warning, so a dangling dataset reference is invisible until `createDataCombined()` fails at run time. A fully static name check is impossible for excel/script sources, but the cheap structural check (non-empty `dataSet` per observed row) plus the removal warning are both feasible.
- [x] R/plots.R:115-209: `.validatePlots()` never checks `plotType` membership in `.validPlotTypes` (only `addPlot()` enforces the enum), so an unknown type coming from JSON passes `validateProject()` and then crashes `createPlots()` with `attempt to set an attribute on NULL` (reproduced), or silently drops the plot from a mixed grid.
- [x] R/plots.R:196-204 vs R/create-plots.R:410-412: severity mismatch. The validator demotes unknown plot-grid `plotIDs` to a warning with the rationale "unknown grid ids fail softly during plot creation", but `createPlots()` hard-errors on exactly that condition via `.validatePlotGridsFromExcel()`. Align the two (and fix the docstring, which is factually wrong).
- [x] R/plots.R:183-206: the plotGrid reference check is skipped entirely whenever `plotConfiguration` is empty, which is exactly the state after removing the last plot (a flow `removePlot()` explicitly warns about).
- [x] R/parameter-identification.R:364-416, 466-483: `.validatePI` does not check that parameter/mapping `scenarios` are a subset of the task's `scenarios`; such configs pass `validateProject()` then hard-fail the runtime build. The runtime is also internally inconsistent about it: step 1 silently skips unknown scenarios while steps 2 and 3 abort.
- [x] R/validation.R:71-73, 98-105: `validationResult$set_data()` and the `data` field are never populated by any adapter (`get_summary()$has_data` is always FALSE) although the documented contract says the result carries the validated data; `validationSummary()` (779-804) still reports `files_with_errors` / `files_with_warnings` though the keys are sections, not files. Implement or trim.
- [x] Scenario `modelFile` existence is never checked by `validateProject()`: the scenarios validator (R/scenarios.R, `.validateScenarios()`) only verifies `modelFile` is a non-empty string, never that the referenced `.pkml` resolves on disk. A scenario pointing at a missing or renamed model file passes `validateProject()` with zero critical errors and no warning, then fails late inside `ospsuite::loadSimulation()` at `runScenarios()` time (R/scenario-execution.R, the `loadSimulation()` call after `modelFilePath` resolution) with an upstream OSPS message that carries no scenario or project context. This is the surviving half of the `must_work` removal tracked in #1031: `.clean_path()` (R/project.R) was deliberately made side-effect-free (no `mustWork` stat) to stop accessor/`print()` reads from emitting `fileNotFound` warnings, which is correct, but nothing took over the existence check it used to do. The fix belongs here, not in `.clean_path()`: add a filesystem-existence check to the scenarios validator that emits a `add_warning("File Not Found", ...)` (or critical error) per scenario whose resolved `modelFile` is absent, mirroring the pattern `.validateObservedData()` already uses for `excel` / `pkml` / `script` entries (R/observed-data.R). Reuse the same path-resolution rule as the runtime (absolute `modelFile` used as-is, relative resolved against `project$modelFolder`). Folder fields (`modelFolder` / `dataFolder` / `configurationsFolder`) lost the same lazy stat; decide whether a missing-but-set folder is worth its own validator warning too.

## Required tests

The suite passes today (1014 tests) while the parameter-set validator validates nothing; that is the headline lesson. **Validation tests must build their fixtures through the parser or the public mutators, never hand-rolled shapes**, otherwise the validator and the data can drift apart again unnoticed. The fix PR must include:

- an invalid parameter set in the real parsed shape that `validateProject()` must flag (critical error), for all three parameter-set sections,
- cross-reference tests for scenario to outputPath and dataCombined to observed dataSet,
- a scenario name containing glue metacharacters (`{`, `}`) driven through `.abortValidationErrors()`,
- an unknown `plotType` fixture that `validateProject()` must flag,
- the currently untested cross-reference branches (grep-verified zero coverage): scenario to `populationId`, scenario to `modelParameterSets`, scenario to `applicationProtocol`, application to `parameterSets`, dataCombined to `scenarios`, and PI outputMapping to `outputPathId`,
- a scenario whose `modelFile` does not resolve on disk (relative against `modelFolder`, and absolute), asserting `validateProject()` surfaces it up front rather than deferring to the `loadSimulation()` abort at run time.

### PR DESCRIPTION
**Pull request by Felixmil**
The chapter 8d audit (#1030) found that `validateProject()` is the gate `runScenarios()` / `createPlots()` / `runPI()` rely on, yet it had holes exactly at the chapter seams: things it waved through that then failed later with worse errors, or never failed at all. This PR closes the validation gaps filed in #1051, building every test fixture through `loadProject()` or the public mutators so the validators and the real parsed data shapes cannot drift apart again unnoticed.

## Parameter-set validation

`.validateParameterSets()` read a `paths` / `values` / `units` shape that neither the parser nor the mutators ever produce, so every set short-circuited and all three parameter-set sections were effectively unvalidated. It now reads the array-of-records shape (`containerPath`, `parameterName`, `value`, `units`) the parser and the `add*ParameterEntry()` family emit, flagging empty parameter paths, non-numeric values, and duplicate paths.

## Glue injection in the abort path

`.abortValidationErrors()` interpolated raw validation messages (which embed user-controlled ids) into `cli` bullets, so a scenario named `Dose {mg}` crashed the abort itself. Messages are now brace-escaped before interpolation so metacharacters render literally.

## Missing cross-reference and section checks

- Scenario to output-path references are now checked in `.validateCrossReferences()`, mirroring the serializer's reverse map; removing a used output path is no longer silently clean only to abort later in `saveProject()`.
- `.validatePlots()` now flags unknown `plotType` values, performs a structural check on empty observed `dataSet` references, and promotes unknown plot-grid `plotID` references from a warning to a critical error, aligning the validator with the hard failure in `createPlots()`. The plot-grid check also runs when `plotConfiguration` is empty (the state after removing the last plot).
- `.validatePI()` now flags `PIParameter` and `PIOutputMapping` scenarios that are not a subset of their task's `scenarios`, the configs that previously passed validation then hard-failed the runtime build.
- `removeObservedData()` now warns when the removed dataset is still referenced by a `dataCombined` observed entry, matching the referenced-elsewhere convention of the other `remove*()` mutators.
- `.validateScenarios()` now checks that each scenario's `modelFile` resolves on disk (relative to `project$modelFolder`), emitting a warning rather than deferring to a hard failure inside `ospsuite::loadSimulation()` at run time.

## Tests

Adds regression coverage for every finding plus the previously uncovered cross-reference branches (scenario to populationId / modelParameterSets / applicationProtocol, application to parameterSets, dataCombined to scenarios, PI outputMapping to outputPathId). The first commit lands the failing tests against the unfixed code; the second lands the fix. The modelFile existence check includes tests for the missing-file warning and the no-warning-when-file-exists case.

Closes #1051

### CODE REVIEW

File: R/validation.R [hunk 1]
```diff
@@ -490,28 +452,23 @@ validateProject <- function(project) {
...
+    )
+    parameterNames <- vapply(
+      set,
+      function(e) as.character(e$parameterName %||% NA_character_),
+      character(1)
+    )
+    values <- lapply(set, function(e) e$value)
 
-    if (any(is.na(paths) | paths == "")) {
+    if (any(is.na(parameterNames) | parameterNames == "")) {
```
Comment (Felixmil) [inline R/validation.R]: **Needs Discussion** — The "empty parameter paths" check only inspects `parameterName`; `containerPath` is never validated for emptiness. The full path is `containerPath|parameterName` (per the doc comment and the `fullPaths` line below), so a record with `containerPath = ""` and a valid `parameterName` produces the malformed path `"|Foo"` and sails through clean.

This is precisely the class of seam-gap this PR sets out to close, so it stands out. If an empty `containerPath` is legitimate for some parameter sets (e.g. top-level/global parameters), this is fine — but then the message wording ("empty parameter paths") is misleading and should say "parameter names". If it isn't legitimate, the condition should also flag empty `containerPaths`:

```suggestion
    if (any(is.na(parameterNames) | parameterNames == "" |
      is.na(containerPaths) | containerPaths == "")) {
```

Either way, confirm the intent — the current code asserts more than it checks.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: R/validation.R [hunk 2]
```diff
@@ -537,7 +499,8 @@ validateProject <- function(project) {
       )
     }
 
-    dupes <- paths[duplicated(paths) & !is.na(paths)]
+    fullPaths <- paste(containerPaths, parameterNames, sep = "|")
+    dupes <- fullPaths[duplicated(fullPaths) & !is.na(fullPaths)]
```
Comment (Felixmil) [inline R/validation.R]: **Noted** (minor) — The `& !is.na(fullPaths)` guard is dead code. `fullPaths` is the result of `paste(containerPaths, parameterNames, sep = "|")`, and `paste()` never yields `NA` — an `NA_character_` input is coerced to the literal string `"NA"`. So `!is.na(fullPaths)` is always `TRUE`.

This was meaningful in the old shape where `dupes <- paths[duplicated(paths) & !is.na(paths)]` operated on a raw vector that could contain `NA`. After the rewrite it's a vestige carried over from the deleted line. Drop it:

```suggestion
    dupes <- fullPaths[duplicated(fullPaths)]
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Felixmil — COMMENTED** [review]
Review feedback assisted by the [critical-code-reviewer skill](https://github.com/posit-dev/skills/blob/main/posit-dev/critical-code-reviewer/SKILL.md).

## Summary

This is a well-constructed PR that does what it claims. The thesis — build every fixture through `loadProject()` / public mutators so the validators are exercised against the real parsed data shapes — is the right call and is executed consistently. The single highest-value change is the `paths`/`values` → array-of-records rewrite in `.validateParameterSets()`: the old shape was never produced by the parser or the mutators, so all three parameter-set sections were silently unvalidated. That's a real, latent hole, now closed and regression-tested against the real shape.

The security-flavored fixes are genuinely careful, not cargo-culted. `.escapeCliBraces()` neutralises user-controlled ids (a scenario named `Dose {mg}`) before they hit `cli`'s glue interpolation, and the snapshot test proves the braces render literally rather than crashing the abort path. The companion fix in `runPI()`'s soft-fail handler — binding the optimizer's `e$message` as a local and interpolating once, instead of routing an uncontrolled string back through a pre-rendered `messages$` template — shows the author understood *why* the double-interpolation was the bug, not just that it crashed.

The validator/runtime alignment work is sound: promoting unknown plot-grid `plotID` refs from warning → critical error (and running the check even when `plotConfiguration` is empty), flagging PI parameter/mapping scenarios outside their task's `scenarios`, and adding the scenario→outputPath cross-reference all converge the validator with the hard failures that previously surfaced later in `createPlots()` / `runPI()`. Moving `.createSinglePITask()` out of the `tryCatch` so build errors propagate while only `task$run()` stays soft is the correct fault boundary.

Verified independently of the description: all 10 CI checks pass; `validateIsOfType` resolves unqualified via `import(ospsuite.utils)` in NAMESPACE (the de-qualification is consistent with the rest of the package); `.validPlotTypes` is defined and matches the enum used by `addPlot()`; and the `when = "6.1.0"` → `"6.0.0"` deprecation edits are a correct consistency fix (package is at 6.0.0.9001 and every other `deprecate_warn` uses 6.0.0).

## Critical Issues (Blocking)

None.

## Required Changes

None blocking. Two items left inline.

## Suggestions

1. **`.validateParameterSets()` empty-path check is narrower than its message** (inline, `R/validation.R:471`). The "empty parameter paths" critical error only inspects `parameterName`; an empty `containerPath` with a valid name yields the malformed full path `"|Foo"` and passes clean. Given this PR's entire purpose is closing seam-gaps, confirm whether empty `containerPath` is legitimate — if not, extend the check; if so, the message wording should be corrected.

2. **Dead `!is.na()` guard** (inline, `R/validation.R:503`). `!is.na(fullPaths)` can never be `FALSE` because `paste()` coerces `NA` to `"NA"`. Vestige from the old raw-vector shape; drop it.

The test suite is the strongest part of this PR — real-shape fixtures, explicit comments where an injection deliberately bypasses a mutator guard to simulate a hand-edited `Project.json`, and coverage for the previously-untested cross-reference branches. Remaining items are minor.

## Verdict

Needs Discussion — no blocking issues; resolve the empty-`containerPath` question (item 1) before this merges as "gaps closed."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: R/validation.R (hunk 1 above)
Comment (Felixmil) [reply R/validation.R]: Fixed in e28fb1c5. Investigation confirmed that an empty `containerPath` is never legitimate (the programmatic mutator already rejects it). The check now also flags empty `containerPath`, and a test covers the previously uncovered case of an empty container path with a valid parameter name.

File: R/validation.R (hunk 2 above)
Comment (Felixmil) [reply R/validation.R]: Removed in e28fb1c5. `paste()` always coerces `NA` to the string `"NA"`, so the guard was indeed unreachable. Dropped.

**Review by rengelke — CHANGES_REQUESTED** [review]
Nice audit follow-up — catching the parameter-set shape mismatch and the glue injection bug in the same pass is a real quality win.

---

**[1]** [`R/parameter-identification.R`:L184](https://github.com/esqLABS/esqlabsR/blob/validation-gaps/R/parameter-identification.R#L184) — maintainability

`PITask()` uses `stop()` for the `outputMappings` element-type check while every other validation in the same function and in `PIParameter()` uses `cli::cli_abort()`. This produces an un-styled plain error message on that one code path.

> 💡 Replace `stop(messages$errorPIWrongElementType(...))` with `cli::cli_abort(messages$errorPIWrongElementType(...))`.

---

**[2]** [`R/project-lifecycle.R`:L83](https://github.com/esqLABS/esqlabsR/blob/validation-gaps/R/project-lifecycle.R#L83) — maintainability

`saveProject()` uses a bare `stop()` with a plain string literal for the missing-path error. No `messages.R` entry, no `cli::cli_abort()`. Two other error paths in the same file follow the same pattern.

> 💡 Add a `messages$missingJsonPath()` entry and replace `stop("No path specified...")` with `cli::cli_abort(messages$missingJsonPath())`.

---

**[3]** [`R/validation.R`:L799](https://github.com/esqLABS/esqlabsR/blob/validation-gaps/R/validation.R#L799) — maintainability

`isAnyCriticalErrors()` and `validationSummary()` are exported functions missing `@title` and `@examples` in their roxygen2 docs. The sibling `validateProject()` in the same file demonstrates the expected pattern.

> 💡 Add `@title`, typed `@param` (`Named list of class "ValidationResults", output from [validateProject()]`), and `@examples \dontrun{...}` blocks to both functions.

---

**[4]** [`R/validation.R`:L349](https://github.com/esqLABS/esqlabsR/blob/validation-gaps/R/validation.R#L349) — correctness

The `"outputPath"` branch in `.warnIfReferenced()` tests `pathValue %in% sc$outputPaths` (a value-based check) instead of `id %in% names(sc$outputPaths)` (id-based, the pattern every other branch uses). If two ids share the same literal path — reachable because `.validateOutputPaths()` only warns — this produces a false positive.

> 💡 Change to `isTRUE(id %in% names(sc$outputPaths))`.

---

**[5]** [`R/project.R`:L319](https://github.com/esqLABS/esqlabsR/blob/validation-gaps/R/project.R#L319) — design

`.markSaved()` resets `validatedSinceMutation` to `FALSE`, which forces re-validation after `saveProject()` even though nothing changed. Saving is not a mutation. The field's own docs confirm it should only be cleared by `.invalidate()` (mutations) and `.read_json()` (load).

> 💡 Remove the `private$.validatedSinceMutation <- FALSE` line from `.markSaved()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Felixmil** [comment]
Thanks for the review. Triaged all five findings.

## Fixed here (commit a832b7be, pushed):
- **[3] roxygen on `isAnyCriticalErrors()` / `validationSummary()`.** Both had a bare one-line roxygen block. Completed them with `@title`, `@description`, a typed `@param` (`Named list of class "ValidationResults"`), `@seealso`, and a `\dontrun{}` `@examples`, matching the sibling `validateProject()`. Also documented `validationSummary()`'s renamed `sections_with_*` return keys. In scope because this PR already edits `validationSummary()` (the key rename carried in). `pkgdown::check_pkgdown()` is clean and the 107 validation tests still pass.

## Deferred (verified out of scope — code this PR does not introduce, already tracked):

### [1] `PITask()` `stop()` vs `cli::cli_abort()` (`parameter-identification.R:184`)
Why deferred: the `PITask()` constructor is unchanged by this PR (the `stop()` is from #1037, ch7). Migrating it is part of the package-wide message-convention sweep, not a one-line edit, since a pre-rendered `messages$` helper passed back into `cli_*()` is re-parsed as a glue template.
Tracked in: #1057 (the bare-`stop()` → `cli_abort(messages$…)` migration surface).

### [2] `saveProject()` bare `stop()` (`project-lifecycle.R:83`)
Why deferred: `project-lifecycle.R` is not in this PR's diff at all, and the fix asks for a net-new `messages.R` entry.
Tracked in: #1057 (verbatim: "project-lifecycle.R:83 inline base `stop()` → `cli_abort(messages$noProjectPath())`").

### [4] `.warnIfReferenced()` outputPath value-based vs id-based (`R/validation.R:349`)
The finding is correct: the `"outputPath"` branch matches `pathValue %in% sc$outputPaths` (value-based) while every other branch, and this PR's own cross-reference check at L670 (`names(sc$outputPaths)`), is id-based. Two ids sharing a literal path would false-positive.
Why deferred: `.warnIfReferenced()` is a `remove*()` mutator helper from #1032; this PR touches `.validateCrossReferences()`, not this branch. It belongs with the `.warnIfReferenced()` consistency work, not the validation-gaps cross-ref additions.
Tracked in: #1055 (the wave-2 `.warnIfReferenced()` / cross-section consistency sweep).

### [5] `.markSaved()` clears `validatedSinceMutation` (`R/project.R:319`)
The finding is correct: saving is not a mutation, so the flag should survive `saveProject()`; the field docs say it is cleared only by mutation (`.invalidate()`) and load (`.read_json()`).
Why deferred: `R/project.R` is not in this PR's diff (the line is from #1044, ch8a), and the `validatedSinceMutation` lifecycle is owned by the `mutation-surface` branch (#1060), where editing `.markSaved()` here would have collided.
Status: fixed on `mutation-surface` (#1060) in commit `9f578def`. `.markSaved()` now clears only `modified`; its roxygen documents that saving leaves `validatedSinceMutation` untouched, and a `test-project.R` case (".markSaved clears modified but leaves validatedSinceMutation set") pins the behavior. It landed where it was routed and stays out of scope for this PR.

Fixing the deferred four here would pre-empt two cross-cutting decisions the audit issues deliberately reserve for single, consistent sweeps (the message-convention migration in #1057 and the mutation-surface invalidation design in #1048/#1055) and would touch files owned by in-flight sibling branches — i.e. scope creep.

**Review by rengelke — APPROVED** [review]
2 Fixed / 3 Deferred (tracked in #1055, #1057)
